File: Winfxklia/app/src/main/java/com/winfxk/winfxklia/dialog/Toast.java
package com.winfxk.winfxklia.dialog;


import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import com.winfxk.winfxklia.R;
import com.winfxk.winfxklia.tool.Tool;

/**
 * @author Winfxk
 */
@SuppressWarnings("unused")
@SuppressLint("InflateParams")
public class Toast {
    public static final int[] Images1 = {
            R.drawable.winfxkliba_toast_icon1, R.drawable.winfxkliba_toast_icon2,
            R.drawable.winfxkliba_toast_icon3, R.drawable.winfxkliba_toast_icon4,
            R.drawable.winfxkliba_toast_icon5, R.drawable.winfxkliba_toast_icon6,
            R.drawable.winfxkliba_toast_icon7, R.drawable.winfxkliba_toast_icon8,
            R.drawable.winfxkliba_toast_icon9, R.drawable.winfxkliba_toast_icon10,
            R.drawable.winfxkliba_toast_icon11, R.drawable.winfxkliba_toast_icon12,
            R.drawable.winfxkliba_toast_icon13, R.drawable.winfxkliba_toast_icon14,
            R.drawable.winfxkliba_toast_icon15, R.drawable.winfxkliba_toast_icon16,
            R.drawable.winfxkliba_toast_icon17, R.drawable.winfxkliba_toast_icon18,
            R.drawable.winfxkliba_toast_icon19, R.drawable.winfxkliba_toast_icon20};
    public static final int[] Images2 = {
            R.drawable.winfxkliba_toast1_icon1, R.drawable.winfxkliba_toast1_icon2,
            R.drawable.winfxkliba_toast1_icon3, R.drawable.winfxkliba_toast1_icon4,
            R.drawable.winfxkliba_toast1_icon5, R.drawable.winfxkliba_toast1_icon6,
            R.drawable.winfxkliba_toast1_icon7, R.drawable.winfxkliba_toast1_icon8,
            R.drawable.winfxkliba_toast1_icon9, R.drawable.winfxkliba_toast1_icon10,
            R.drawable.winfxkliba_toast1_icon11, R.drawable.winfxkliba_toast1_icon12,
            R.drawable.winfxkliba_toast1_icon13, R.drawable.winfxkliba_toast1_icon14,
            R.drawable.winfxkliba_toast1_icon15, R.drawable.winfxkliba_toast1_icon16,
            R.drawable.winfxkliba_toast1_icon17, R.drawable.winfxkliba_toast1_icon18,
            R.drawable.winfxkliba_toast1_icon19, R.drawable.winfxkliba_toast1_icon20};
    public static final int LENGTH_SHORT = 0;
    public static final int LENGTH_LONG = 1;
    private ImageView imageView, imageView1;
    private final Context activity;
    private TextView textView;
    private int showTime;
    private View view;

    public Toast(Context context) {
        this.activity = context;
        LayoutInflater inflate = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        view = inflate.inflate(R.layout.winfxkliba_toast, null);
        textView = view.findViewById(R.id.textView1);
        imageView = view.findViewById(R.id.imageView1);
        imageView1 = view.findViewById(R.id.imageView2);
    }

    /**
     * 显示文本
     */
    public void show() {
        android.widget.Toast toast = new android.widget.Toast(activity);
        toast.setView(view);
        toast.setDuration(showTime);
        toast.show();
    }

    /**
     * 获取显示的图片视图1的对象
     */
    public ImageView getImageView() {
        return imageView;
    }

    /**
     * 获取显示的图片视图2的对象
     *
     * @return
     */
    public ImageView getImageView1() {
        return imageView1;
    }

    /**
     * 获取显示的文本视图对象
     *
     * @return
     */
    public TextView getTextView() {
        return textView;
    }

    /**
     * 获取显示的时长
     *
     * @return
     */
    public int getTime() {
        return showTime;
    }

    /**
     * 获取显示的视图
     *
     * @return
     */
    public View getView() {
        return view;
    }

    /**
     * 设置显示的文本视图
     *
     * @param textView
     */
    public void setTextView(TextView textView) {
        this.textView = textView;
    }

    /**
     * 设置显示的图片1
     *
     * @param imageView
     */
    public void setImageView(ImageView imageView) {
        this.imageView = imageView;
    }

    /**
     * 设置显示的图片2
     *
     * @param imageView1
     */
    public void setImageView1(ImageView imageView1) {
        this.imageView1 = imageView1;
    }

    /**
     * 设置显示的视图
     *
     * @param view
     */
    public void setView(View view) {
        this.view = view;
    }

    /**
     * 设置显示的时长
     *
     * @param i
     */
    public void setTime(int i) {
        this.showTime = i;
    }

    /**
     * 设置显示的文本
     *
     * @param string
     */
    public void setText(String string) {
        textView.setText(string);
    }

    /**
     * 设置显示的图片
     *
     * @param drawable
     */
    public void setImage(Drawable drawable) {
        imageView.setImageDrawable(drawable);
        imageView1.setImageDrawable(drawable);
    }

    /**
     * 设置显示的图片
     *
     * @param bm
     */
    public void setImage(Bitmap bm) {
        imageView.setImageBitmap(bm);
        imageView1.setImageBitmap(bm);
    }

    /**
     * 设置显示的图片
     *
     * @param resId
     */
    public void setImage(int resId) {
        imageView.setImageResource(resId);
        imageView1.setImageResource(resId);
    }

    /**
     * 设置显示的图片
     *
     * @param uri
     */
    public void setImage(Uri uri) {
        imageView.setImageURI(uri);
        imageView1.setImageURI(uri);
    }

    /**
     * 快捷提示一个窗口
     *
     * @param context Activity对象
     * @param string  要提示的内容
     * @param i       <b>提示时间</b>
     * @return
     * @see #LENGTH_SHORT
     * @see #LENGTH_LONG
     */
    public static Toast makeText(Context context, Object string, int i) {
        Toast tsate = new Toast(context);
        tsate.setText(String.valueOf(string));
        tsate.setTime(i);
        tsate.setRandImage();
        return tsate;
    }

    /**
     * 快捷显示一个提示
     *
     * @param context Activity对象
     * @param string  要显示的文本
     * @return
     */
    public static Toast makeText(Context context, Object string) {
        return makeText(context, String.valueOf(string), LENGTH_SHORT);
    }

    /**
     * 设置随机字体颜色
     *
     * @param r
     * @param g
     * @param b
     */
    public void setTextColor(int r, int g, int b) {
        GradientDrawable drawable = (GradientDrawable) textView.getBackground();
        drawable.setColor(Color.rgb(255 - r, 255 - g, 255 - b));
        textView.setBackground(drawable);
        textView.setTextColor(Color.rgb(r, g, b));
    }

    /**
     * 设置随机字体颜色
     */
    public void setRandColor() {
        setTextColor(Tool.getRand(0, 255), Tool.getRand(0, 255), Tool.getRand(0, 255));
    }

    /**
     * 设置随机字体背景颜色
     *
     * @param r
     * @param g
     * @param b
     */
    public void setTextBackgColor(int r, int g, int b) {
        GradientDrawable drawable = (GradientDrawable) textView.getBackground();
        drawable.setColor(Color.rgb(r, g, b));
        textView.setBackground(drawable);
        textView.setTextColor(Color.rgb(255 - r, 255 - g, 255 - b));
    }

    /**
     * 设置随机字体背景颜色
     */
    public void setRandBackgColor() {
        setTextBackgColor(Tool.getRand(0, 255), Tool.getRand(0, 255), Tool.getRand(0, 255));
    }

    /**
     * 设置随机图片跟随
     */
    public void setRandImage() {
        int[] Images1 = Tool.getRand(1, 2) == 1 ? Toast.Images1 : Images2;
        int r = Images1[Tool.getRand(0, Images1.length - 1)];
        int r2 = Images1[Tool.getRand(0, Images1.length - 1)];
        while (r == r2) r2 = Images1[Tool.getRand(0, Images1.length - 1)];
        getImageView().setImageResource(r);
        getImageView1().setImageResource(r2);
    }
}

File: mylibrary/src/main/java/cn/winfxk/android/mylibrary/tip/Toast.kt
/*
* Copyright Notice
* © [2024 - 2026] Winfxk. All rights reserved.
* The software, its source code, and all related documentation are the intellectual property of Winfxk. Any reproduction or distribution of this software or any part thereof must be clearly attributed to Winfxk and the original author. Unauthorized copying, reproduction, or distribution without proper attribution is strictly prohibited.
* For inquiries, support, or to request permission for use, please contact us at:
* Email: admin@winfxk.cn
* QQ: 2508543202
* Visit our homepage for more information: http://Winfxk.cn
*
* --------- Create message ---------
* Created by AI Assistant
* Author： Winfxk
* Web: http://winfxk.com
* Created Date: 2026/06/03 09:32
*/
package cn.winfxk.android.mylibrary.tip

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.ContextWrapper
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import cn.winfxk.android.mylibrary.R
import java.lang.ref.WeakReference
import java.util.LinkedList

@Suppress("unused")
@SuppressLint("InflateParams")
class Toast private constructor(private val context: Context) {
    private val view: View by lazy { LayoutInflater.from(context).inflate(R.layout.winfxklia_toast, null) }
    private val imageView: ImageView by lazy { view.findViewById(R.id.imageView1) }
    private val imageView1: ImageView by lazy { view.findViewById(R.id.imageView2) }
    private val textView: TextView by lazy { view.findViewById(R.id.textView1) }
    var time: Int = LENGTH_SHORT
    private var task: ToastTask? = null

    init {
        LifecycleTracker.init(context)
    }

    /**
     * 显示 Toast
     */
    fun show() {
        val targetActivity = getActivity(context) ?: LifecycleTracker.currentActivity
        if (targetActivity == null) {
            Log.w("WinfxkToast", "当前不在前台 Activity 环境中，回退至原生 Toast")
            android.widget.Toast.makeText(context, textView.text, time).show()
            return
        }
        val durationMs = if (time == LENGTH_LONG) 3500L else 2000L
        task = ToastTask(WeakReference(targetActivity), view, durationMs)
        ToastManager.enqueue(task !!)
    }

    /**
     * 提前取消显示
     */
    fun cancel() {
        task?.let { ToastManager.cancel(it) }
    }

    fun setText(string: String?) {
        textView.text = string
    }

    fun setImage(drawable: Drawable?) {
        imageView.setImageDrawable(drawable)
        imageView1.setImageDrawable(drawable)
    }

    fun setImage(bm: Bitmap?) {
        imageView.setImageBitmap(bm)
        imageView1.setImageBitmap(bm)
    }

    fun setImage(resId: Int) {
        imageView.setImageResource(resId)
        imageView1.setImageResource(resId)
    }

    fun setImage(uri: Uri?) {
        imageView.setImageURI(uri)
        imageView1.setImageURI(uri)
    }

    fun setTextColor(r: Int, g: Int, b: Int) {
        val drawable = textView.background as GradientDrawable
        drawable.setColor(Color.rgb(255 - r, 255 - g, 255 - b))
        textView.background = drawable
        textView.setTextColor(Color.rgb(r, g, b))
    }

    fun setRandColor() {
        setTextColor((0 .. 255).random(), (0 .. 255).random(), (0 .. 255).random())
    }

    fun setTextBackgColor(r: Int, g: Int, b: Int) {
        val drawable = textView.background as GradientDrawable
        drawable.setColor(Color.rgb(r, g, b))
        textView.background = drawable
        textView.setTextColor(Color.rgb(255 - r, 255 - g, 255 - b))
    }

    fun setRandBackgColor() {
        setTextBackgColor((0 .. 255).random(), (0 .. 255).random(), (0 .. 255).random())
    }

    fun setRandImage() {
        val r = Images1.random()
        var r2 = Images1.random()
        while (r == r2) r2 = Images1.random()
        this.imageView.setImageResource(r)
        this.imageView1.setImageResource(r2)
    }

    private fun getActivity(context: Context?): Activity? {
        return when (context) {
            is Activity       -> context
            is ContextWrapper -> getActivity(context.baseContext)
            else              -> null
        }
    }

    companion object {
        internal val Images1: IntArray = intArrayOf(
            R.drawable.winfxklia_toast_icon1, R.drawable.winfxklia_toast_icon2,
            R.drawable.winfxklia_toast_icon3, R.drawable.winfxklia_toast_icon4,
            R.drawable.winfxklia_toast_icon5, R.drawable.winfxklia_toast_icon6,
            R.drawable.winfxklia_toast_icon7, R.drawable.winfxklia_toast_icon8,
            R.drawable.winfxklia_toast_icon9, R.drawable.winfxklia_toast_icon10,
            R.drawable.winfxklia_toast_icon11, R.drawable.winfxklia_toast_icon12,
            R.drawable.winfxklia_toast_icon13, R.drawable.winfxklia_toast_icon14,
            R.drawable.winfxklia_toast_icon15, R.drawable.winfxklia_toast_icon16,
            R.drawable.winfxklia_toast_icon17, R.drawable.winfxklia_toast_icon18,
            R.drawable.winfxklia_toast_icon19, R.drawable.winfxklia_toast_icon20
        )
        const val LENGTH_SHORT: Int = 0
        const val LENGTH_LONG: Int = 1

        @JvmStatic
        fun makeText(context: Context, string: Any?, duration: Int = LENGTH_SHORT): Toast {
            val toast = Toast(context)
            toast.setText(string.toString())
            toast.time = duration
            toast.setRandImage()
            return toast
        }
    }
}

/**
 * 内部任务类，持有弱引用防止内存泄漏
 */
private data class ToastTask(
    val activityRef: WeakReference<Activity>,
    val view: View,
    val durationMs: Long
)

/**
 * 智能队列管理器 (单例)
 */
@SuppressLint("StaticFieldLeak")
private object ToastManager {
    private val queue = LinkedList<ToastTask>()
    private var isShowing = false
    private val handler = Handler(Looper.getMainLooper())
    private var currentTask: ToastTask? = null
    private val dismissRunnable = Runnable { dismissCurrent() }

    fun enqueue(task: ToastTask) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            handler.post { enqueue(task) }
            return
        }
        queue.offer(task)
        showNextIfNeeded()
    }

    fun cancel(task: ToastTask) {
        if (currentTask == task) {
            handler.removeCallbacks(dismissRunnable)
            dismissCurrent()
        } else queue.remove(task)
    }

    private fun showNextIfNeeded() {
        if (isShowing || queue.isEmpty()) return
        val task = queue.poll() ?: return
        val activity = task.activityRef.get()
        if (activity == null || activity.isFinishing || activity.isDestroyed) {
            showNextIfNeeded()
            return
        }
        isShowing = true
        currentTask = task
        val rootView = activity.findViewById<FrameLayout>(android.R.id.content)
        val view = task.view
        (view.parent as? ViewGroup)?.removeView(view)
        val bottomMarginPx = (100 * activity.resources.displayMetrics.density).toInt()
        val params = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.WRAP_CONTENT,
            FrameLayout.LayoutParams.WRAP_CONTENT
        ).apply {
            gravity = Gravity.CENTER_HORIZONTAL or Gravity.BOTTOM
            bottomMargin = bottomMarginPx
        }
        view.alpha = 0f
        view.translationY = 50f
        rootView.addView(view, params)
        view.animate().alpha(1f).translationY(0f).setDuration(300)
            .setInterpolator(DecelerateInterpolator()).withEndAction {
                if (currentTask == task) {
                    handler.removeCallbacks(dismissRunnable)
                    handler.postDelayed(dismissRunnable, task.durationMs)
                }
            }.start()
        view.addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {
            override fun onViewAttachedToWindow(v: View) {}
            override fun onViewDetachedFromWindow(v: View) {
                handler.removeCallbacks(dismissRunnable)
                view.removeOnAttachStateChangeListener(this)
                if (currentTask == task) {
                    currentTask = null
                    isShowing = false
                    showNextIfNeeded()
                }
            }
        })
    }

    private fun dismissCurrent() {
        val task = currentTask ?: return
        val view = task.view
        view.animate().alpha(0f).translationY(50f).setDuration(250)
            .setInterpolator(AccelerateInterpolator()).withEndAction {
                if (currentTask == task) {
                    currentTask = null
                    isShowing = false
                    (view.parent as? ViewGroup)?.removeView(view)
                    showNextIfNeeded()
                } else (view.parent as? ViewGroup)?.removeView(view)
            }.start()
    }
}

/**
 * 自动生命周期追踪器
 */
@SuppressLint("StaticFieldLeak")
private object LifecycleTracker : Application.ActivityLifecycleCallbacks {
    private var isInitialized = false
    private var activityReference: WeakReference<Activity>? = null
    val currentActivity: Activity? get() = activityReference?.get()

    fun init(context: Context) {
        if (isInitialized) return
        val app = context.applicationContext as? Application ?: return
        app.registerActivityLifecycleCallbacks(this)
        isInitialized = true
    }

    override fun onActivityResumed(activity: Activity) {
        activityReference = WeakReference(activity)
    }

    override fun onActivityPaused(activity: Activity) {
        if (currentActivity == activity) activityReference = null
    }

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
    override fun onActivityStarted(activity: Activity) {}
    override fun onActivityStopped(activity: Activity) {}
    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
    override fun onActivityDestroyed(activity: Activity) {}
}